cluster-autoscaler/cloudprovider/gce/mig_info_provider.go

```diff
@@ -0,0 +1,211 @@
+	// GetMigBasename returns basename for given MIG ref
+	GetMigBasename(migRef GceRef) (string, error)
+	// GetMigInstanceTemplateName returns instance template name for given MIG ref
+	GetMigInstanceTemplateName(migRef GceRef) (string, error)
+	// GetMigInstanceTemplate returns instance template for given MIG ref
+	GetMigInstanceTemplate(migRef GceRef) (*gce.InstanceTemplate, error)
+}
+
+type cachingMigInfoProvider struct {
+	mutex     sync.Mutex
+	cache     *GceCache
+	gceClient AutoscalingGceClient
+	projectId string
+}
+
+// NewCachingMigInfoProvider creates an instance of caching MigInfoProvider
+func NewCachingMigInfoProvider(cache *GceCache, gceClient AutoscalingGceClient, projectId string) MigInfoProvider {
+	return &cachingMigInfoProvider{
+		cache:     cache,
+		gceClient: gceClient,
+		projectId: projectId,
+	}
+}
+
+func (c *cachingMigInfoProvider) GetMigTargetSize(migRef GceRef) (int64, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	targetSize, found := c.cache.GetMigTargetSize(migRef)
+	if found {
+		return targetSize, nil
+	}
+
+	err := c.fillMigInfoCache()
+	targetSize, found = c.cache.GetMigTargetSize(migRef)
+	if err == nil && found {
+		return targetSize, nil
+	}
+
+	// fallback to querying for single mig
+	targetSize, err = c.gceClient.FetchMigTargetSize(migRef)
+	if err != nil {
+		return 0, err
+	}
+	c.cache.SetMigTargetSize(migRef, targetSize)
+	return targetSize, nil
+}
+
+func (c *cachingMigInfoProvider) GetMigBasename(migRef GceRef) (string, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	basename, found := c.cache.GetMigBasename(migRef)
+	if found {
+		return basename, nil
+	}
+
+	err := c.fillMigInfoCache()
+	basename, found = c.cache.GetMigBasename(migRef)
+	if err == nil && found {
+		return basename, nil
+	}
+
+	// fallback to querying for single mig
+	basename, err = c.gceClient.FetchMigBasename(migRef)
+	if err != nil {
+		return "", err
+	}
+	c.cache.SetMigBasename(migRef, basename)
+	return basename, nil
+}
+
+func (c *cachingMigInfoProvider) GetMigInstanceTemplateName(migRef GceRef) (string, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	templateName, found := c.cache.GetMigInstanceTemplateName(migRef)
+	if found {
+		return templateName, nil
+	}
+
+	err := c.fillMigInfoCache()
+	templateName, found = c.cache.GetMigInstanceTemplateName(migRef)
+	if err == nil && found {
+		return templateName, nil
+	}
+
+	// fallback to querying for single mig
+	templateName, err = c.gceClient.FetchMigTemplateName(migRef)
+	if err != nil {
+		return "", err
+	}
+	c.cache.SetMigInstanceTemplateName(migRef, templateName)
+	return templateName, nil
+}
+
+func (c *cachingMigInfoProvider) GetMigInstanceTemplate(migRef GceRef) (*gce.InstanceTemplate, error) {
+	templateName, err := c.GetMigInstanceTemplateName(migRef)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	template, found := c.cache.GetMigInstanceTemplate(migRef)
+	if found && template.Name == templateName {
```

Comment:
Maybe it's worth logging if instance template has changed? This should be a very rare event (so no impact on log volume) and it may be quite useful when debugging (ex. see which template was used in particular scale-up).

Reply:
Done